### PR TITLE
Netatmo: avoid remaining in INITIALIZING state in case of API timeout

### LINKED ISSUE #3258: Netatmo: partnerdevices connection timeout => thing remaining in INITIALIZING state.
```
---- ERROR https://api.netatmo.net/api/partnerdevices
java.io.IOException: org.apache.oltu.oauth2.common.exception.OAuthSystemException: java.net.ConnectException: Connection timed out: connect
	at io.swagger.client.auth.OAuth.updateAccessToken(OAuth.java:140)
	at io.swagger.client.auth.OAuth.intercept(OAuth.java:86)
	at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:225)
	at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:199)
	at com.squareup.okhttp.Call.execute(Call.java:79)
	at retrofit.client.OkClient.execute(OkClient.java:53)
	at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:326)
	at retrofit.RestAdapter$RestHandler.invoke(RestAdapter.java:240)
	at com.sun.proxy.$Proxy88.partnerdevices(Unknown Source)
	at org.openhab.binding.netatmo.handler.NetatmoBridgeHandler.lambda$0(NetatmoBridgeHandler.java:121)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.oltu.oauth2.common.exception.OAuthSystemException: java.net.ConnectException: Connection timed out: connect
	at io.swagger.client.auth.OAuthOkHttpClient.execute(OAuthOkHttpClient.java:61)
	at org.apache.oltu.oauth2.client.OAuthClient.accessToken(OAuthClient.java:65)
	at org.apache.oltu.oauth2.client.OAuthClient.accessToken(OAuthClient.java:55)
	at org.apache.oltu.oauth2.client.OAuthClient.accessToken(OAuthClient.java:71)
	at io.swagger.client.auth.OAuth.updateAccessToken(OAuth.java:129)
	... 16 more
Caused by: java.net.ConnectException: Connection timed out: connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at com.squareup.okhttp.internal.Platform.connectSocket(Platform.java:107)
	at com.squareup.okhttp.Connection.connect(Connection.java:156)
	at com.squareup.okhttp.Connection.connectAndSetOwner(Connection.java:175)
	at com.squareup.okhttp.OkHttpClient$1.connectAndSetOwner(OkHttpClient.java:120)
	at com.squareup.okhttp.internal.http.HttpEngine.nextConnection(HttpEngine.java:330)
	at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:319)
	at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:241)
	at com.squareup.okhttp.Call.getResponse(Call.java:271)
	at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:228)
	at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:199)
	at com.squareup.okhttp.Call.execute(Call.java:79)
	at io.swagger.client.auth.OAuthOkHttpClient.execute(OAuthOkHttpClient.java:54)
	... 20 more

---- END ERROR
2018-02-18 14:17:43.732 [WARN ] [rappedScheduledExecutorService:67   ] - Scheduled runnable ended with an exceptionjava.lang.NullPointerException: null
	at org.openhab.binding.netatmo.handler.NetatmoBridgeHandler.lambda$0(NetatmoBridgeHandler.java:126)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

It leads to a NPE in `scheduleTokenInitAndRefresh` in the catch code and so this code is not rescheduled. Last consequence is that the brigde thing remains in a INITIALIZING state.

As a side note, I don't understand why the API exits now in timeout, it was working well 2 hjours ago !

By the way, the issue is that we should catch this timeout exception and put the thing in a OFFLINE state.

### PR DESCRIPTION
**Pull request by lolodomo**
Fixes #3258

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
Yes, unfortunately. The binding is no more working for me since yesterday afternoon  due to an API call exiting in timeout. Without this fix, the thing state remains in INITIALIZING state.

I hope that my account was just banished for one day. Can you please restart ypur OH and check that everything is ok for you ?

**Comment by lolodomo** [comment]
Good news for me, with this proposed fix run in my production environment, the binding was able to retry connection every 90 minutes and finally my bridge thing was ONLINE when I came home this evening. Strangely, my weather station thing was ONLINE too but my external module thing was still OFFLINE, that is not the expected result.

I think I was banished during 24 hours after probably too much usage of the API due to the issue #3256.

So at least, this fix is working well ;)

**Comment by cweitkamp** [comment]
Maybe I'm too tired for tonight but honestly I didn't see the code which causes the NPE mentioned in #3258. Would you be so kind and explain in detail how your fix solves this? TYVM.

File: addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java [hunk 1]
```diff
@@ -119,23 +120,32 @@ private void scheduleTokenInitAndRefresh() {
...
-                    default:
-                        logger.error("Unable to connect Netatmo API : {}", e.getMessage(), e);
-                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                                "Unable to connect Netatmo API : " + e.getLocalizedMessage());
-                        return;
+                if (e.getKind() == Kind.NETWORK) {
+                    logger.warn("Network error while connecting to Netatmo API, will retry in {} s",
+                            configuration.reconnectInterval);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR,
+                            "Netatmo Access Failed, will retry in " + configuration.reconnectInterval.toString()
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java]: Minor: the `toString` is / was not needed here

File: addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java [hunk 2]
```diff
@@ -119,23 +120,32 @@ private void scheduleTokenInitAndRefresh() {
...
+                    switch (e.getResponse().getStatus()) {
+                        case 404: // If no partner station has been associated - likely to happen - we'll have this
+                                  // error
+                                  // but it means connection to API is OK
+                            connectionSucceed();
+                            break;
+                        case 403: // Forbidden Access maybe too many requests ? Let's wait next cycle
+                            logger.warn("Error 403 while connecting to Netatmo API, will retry in {} s",
+                                    configuration.reconnectInterval);
+                            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_REGISTERING_ERROR,
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java]: Are you sure these thingstatusDetail are correct, @htreu wdyt?

**Comment by lolodomo** [comment]
`e.getResponse().getStatus()` was triggering the NPE I think.

File: addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java (hunk 2 above)
Comment (lolodomo) [reply addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java]: Note that this part of code is not mine. That's true that I was a little surprised to see that detailed status.

Comment (htreu) [reply addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoBridgeHandler.java]: Unfortunately `ThingStatusDetail.HANDLER_REGISTERING_ERROR` has no JavaDoc but from the context it seems to be wrong here. I suggest to go with `ThingStatusDetail .COMMUNICATION_ERROR` along with the already provided description.

**Pull request merged**